File: api/line/line_signature.py
"""LINE Messaging API Webhookの署名を検証する。"""

from __future__ import annotations

import base64
import hashlib
import hmac
import os
from pathlib import Path

from dotenv import load_dotenv


PROJECT_ROOT = Path(__file__).resolve().parents[2]
load_dotenv(PROJECT_ROOT / "config" / ".env")


def verify_line_signature(body: bytes, signature: str) -> bool:
    """Webhook本文とX-Line-Signatureが正しいか検証する。"""
    channel_secret = os.getenv("LINE_CHANNEL_SECRET", "").strip()
    if not body or not signature or not channel_secret:
        return False

    digest = hmac.new(
        channel_secret.encode("utf-8"),
        body,
        hashlib.sha256,
    ).digest()
    expected = base64.b64encode(digest).decode("utf-8")
    return hmac.compare_digest(expected, signature)

File: discord-bot/src/integrations/lineSignature.js
import crypto from "node:crypto";

export function verifyLineSignature(config, rawBody, signature) {
  if (!config.line.channelSecret || !signature) {
    return false;
  }

  const expected = crypto
    .createHmac("sha256", config.line.channelSecret)
    .update(rawBody)
    .digest("base64");

  const expectedBuffer = Buffer.from(expected);
  const signatureBuffer = Buffer.from(signature);
  if (expectedBuffer.length !== signatureBuffer.length) {
    return false;
  }

  return crypto.timingSafeEqual(expectedBuffer, signatureBuffer);
}
